ql/src/java/org/apache/hadoop/hive/ql/stats/BasicStatsNoJobTask.java

```diff
@@ -223,6 +227,16 @@ public void run() {
         } else {
           fileList = HiveStatsUtils.getFileStatusRecurse(dir, -1, fs);
         }
+        ThreadPoolExecutor tpE = null;
+        ArrayList<Future<FileStats>> futures = null;
+        int numThreads = HiveConf.getIntVar(jc, HiveConf.ConfVars.BASICSTATSTASKSMAXTHREADS);
+        if (fileList.size() > 1 && numThreads > 1) {
+          numThreads = Math.max(fileList.size(), numThreads);
```

Comment:
Limit to 2*available processors? (i.e instead of fileList.size(). In case file listing is 1k, it shouldn't spin up 1k threads).